### PR TITLE
186 new notebook services api

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
This PR implements the new notebook API as discussed in https://github.com/SwissDataScienceCenter/renku-notebooks/issues/186. A change with that issue is that this implementation replaces `<server_id>` with `<server_name>` which has the same exact value (16-digit hash).

### CODE REVIEW

File: renku_notebooks/util/kubernetes_.py [hunk 1]
```diff
@@ -63,25 +64,35 @@
...
-            current_app.logger.debug(server)
-            return server
-    return {}
+def get_user_servers(user, namespace=None, project=None, branch=None, commit_sha=None):
+    """Fetch all the user named servers that matches the provided criteria"""
+
+    def filter_server(server):
+        def check_annotation(annotation, value):
+            annotation_name = config.RENKU_ANNOTATION_PREFIX + annotation
+            return not value or server["annotations"].get(annotation_name) == value
```
Comment (rokroskar) [inline renku_notebooks/util/kubernetes_.py]: A bit difficult to read the logic here - what about something like

```
if value:
    return server.get("annotations", {}).get(annotation_name, "") == value
return True
```

File: renku_notebooks/util/kubernetes_.py [hunk 2]
```diff
@@ -63,25 +64,35 @@
...
+            check_annotation("namespace", namespace)
+            and check_annotation("projectName", project)
+            and check_annotation("branch", branch)
+            and check_annotation("commit-sha", commit_sha)
+        )
+
+    servers = _get_all_user_servers(user)
+    filtered_servers = [s for s in servers if filter_server(s)]
+    current_app.logger.debug(filtered_servers)
+    return {s["name"]: s for s in filtered_servers}
```
Comment (rokroskar) [inline renku_notebooks/util/kubernetes_.py]: why return just the names and not the full server JSON?

File: renku_notebooks/util/kubernetes_.py [hunk 3]
```diff
@@ -140,8 +151,8 @@ def get_server_url(pod):
 
     pods = get_user_server_pods(user)
 
-    servers = {
-        pod.metadata.annotations["hub.jupyter.org/servername"]: {
+    servers = [
+        {
```
Comment (rokroskar) [inline renku_notebooks/util/kubernetes_.py]: use the name as key - then in `get_user_server` you can just extract it immediately.

File: renku_notebooks/util/jupyterhub_.py [hunk 4]
```diff
@@ -45,3 +50,40 @@ def notebook_url(user, server_name, notebook=None):
...
+    global __auth, __prefix, __headers
+    __auth = auth
+    __prefix = auth.api_url
+    __headers = {auth.auth_header_name: f"token {auth.api_token}"}
+
+
+def check_user_has_named_servers(user, server_name):
+    """Check if the named-server exists in user's JupyterHub servers"""
+    user_info = get_user_info(user)
+    servers = user_info.get("servers")
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku_notebooks/util/jupyterhub_.py]: I remember that `servers` can be `None` at some point. I'll simplify the logic with
```python
return servers is not None and server_name in servers
```

[the remark this answers is not in the document]
Comment (rokroskar) [reply renku_notebooks/util/jupyterhub_.py]: That's why in my suggestion I put `{} ` as a default value to `get`

[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku_notebooks/util/jupyterhub_.py]: I meant `servers` can be `None` in `user_info`, i.e. `{..., servers: null, ...}`. So, when we do `user_info.get("servers", {})` it returns `None` (`get` returns the provided default value only if no value if defined for the requested key e.g. `type({'a': None}.get('a', {}))` returns `None` not `{}`).

[the remark this answers is not in the document]
Comment (rokroskar) [reply renku_notebooks/util/jupyterhub_.py]: Ah I misunderstood - makes sense!

File: renku_notebooks/util/kubernetes_.py (hunk 2 above)
Comment (mohammad-sdsc) [reply renku_notebooks/util/kubernetes_.py]: This returns a map of {server_name: server}. 
I've changed the code here according to your other comment and this is a bit simpler now.

File: renku_notebooks/util/jupyterhub_.py [hunk 5]
```diff
@@ -45,3 +50,40 @@ def notebook_url(user, server_name, notebook=None):
     if notebook:
         notebook_url += "lab/tree/{notebook}".format(notebook=notebook)
     return notebook_url
+
+
+def set_auth(auth):
+    global __auth, __prefix, __headers
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku_notebooks/util/jupyterhub_.py]: I'll move `auth` creation to `jupyterhub_.py` then. Like this, `auth.py` has no direct dependency on `jupyterhub` and all `jupyterhub` required functionality is encapsulated in `jupyterhub_.py`.

File: renku_notebooks/api/notebooks.py [hunk 6]
```diff
@@ -136,87 +138,67 @@ def launch_notebook(user, namespace, project, commit_sha, notebook=None):
...
-            prefix=auth.api_url, user=user, server_name=server_name
-        ),
-        headers=headers,
-    )
+    r = delete_named_server(user, server_name)
     return current_app.response_class(r.content, status=r.status_code)
 
 
-@bp.route("servers/<server_name>", methods=["DELETE"])
+@bp.route("server_options/<server_name>")
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku_notebooks/api/notebooks.py]: My bad! I removed the `<server_name>` part.

File: renku_notebooks/util/kubernetes_.py (hunk 2 above)
Comment (rokroskar) [reply renku_notebooks/util/kubernetes_.py]: argh yes of course. apologies for the stupid question.

**Comment by rokroskar** [comment]
This needs to be coordinated with https://github.com/SwissDataScienceCenter/renku-ui/issues/494 - however, the master branches are not compatible atm anyway.

**Comment by lorenzo-cavazzi** [comment]
Can I "Update branch" to test it with SwissDataScienceCenter/renku-ui#558 ?

**Comment by rokroskar** [comment]
> Can I "Update branch" to test it with [SwissDataScienceCenter/renku-ui#558](https://github.com/SwissDataScienceCenter/renku-ui/pull/558) ?

done

**Pull request merged**